packages/@lwc/engine-core/src/patches/detect-synthetic-cross-root-aria.ts

```diff
@@ -0,0 +1,183 @@
+                const targetRoot = target.getRootNode();
+                if (targetRoot !== root) {
+                    // target element's shadow root is not the same as ours
+                    reportViolation(elm, target, attrName);
+                }
+            }
+        }
+    }
+}
+
+let enabled = false;
+
+// We want to avoid patching globals whenever possible, so this should be tree-shaken out in prod-mode and if
+// reporting is not enabled. It should also only run once
+function enableDetection() {
+    if (enabled) {
+        return; // don't double-apply the patches
+    }
+    enabled = true;
+
+    const { setAttribute } = Element.prototype;
+
+    // Detect calling `setAttribute` to set an idref or an id
+    assign(Element.prototype, {
+        setAttribute(this: Element, attrName: string, attrValue: any) {
+            setAttribute.call(this, attrName, attrValue);
+            if (attrName === 'id' || ID_REFERENCING_ATTRIBUTES_SET.has(attrName)) {
+                detectSyntheticCrossRootAria(this, attrName, attrValue);
+            }
+        },
+    } as Pick<Element, 'setAttribute'>);
+
+    // Detect `elm.id = 'foo'`
+    const idDescriptor = getOwnPropertyDescriptor(Element.prototype, 'id');
+    if (!isUndefined(idDescriptor)) {
+        const { get, set } = idDescriptor;
+        // These should always be a getter and a setter, but if someone is monkeying with the global descriptor, ignore it
+        if (isFunction(get) && isFunction(set)) {
+            defineProperty(Element.prototype, 'id', {
+                get() {
+                    return get.call(this);
+                },
+                set(value: any) {
+                    set.call(this, value);
+                    detectSyntheticCrossRootAria(this, 'id', value);
+                },
+                // On the default descriptor for 'id', enumerable and configurable are true
+                enumerable: true,
+                configurable: true,
+            });
+        }
+    }
+}
+
+// Our detection logic relies on some modern browser features. We can just skip reporting the data
+// for unsupported browsers
+function supportsCssEscape() {
+    return typeof CSS !== 'undefined' && isFunction(CSS.escape);
+}
+
+// If this page is not using synthetic shadow, then we don't need to install detection. Note
+// that we are assuming synthetic shadow is loaded before LWC.
+function isSyntheticShadowLoaded() {
+    // We should probably be calling `renderer.isSyntheticShadowDefined`, but 1) we don't have access to the renderer,
+    // and 2) this code needs to run in @lwc/engine-core, so it can access `logWarn()` and `report()`.
+    return hasOwnProperty.call(Element.prototype, KEY__SHADOW_TOKEN);
+}
+
+// Detecting cross-root ARIA in synthetic shadow only makes sense for the browser
+if (process.env.IS_BROWSER && supportsCssEscape() && isSyntheticShadowLoaded()) {
+    // Always run detection in dev mode, so we can at least print to the console
```

Comment:
In a scenario like the CPM Record Home Inline Edit, with hundreds of violations, or other similar scenarios, how noisy/scary does the console get if we print every single violation? (are they generally the same message collapsed together or is there quite a variety of combinations?)

We might not be able to help it, but I know one occasional pain point with dev productivity has been dealing with a large amount of unhelpful/irrelevant logs/warnings in the console (it's gotten better over the years though). We should consider whether the potential productivity/psychological impact on those who can't do anything about the warning (besides file bugs on other teams) is what we're okay with too.

Reply:
@jye-sf That's a good point. We can coalesce based on the tagName or something and only log once, rather than 1,000s of times.

Or alternatively, we can get rid of the console warning altogether. Right now the important thing is the reporting.

As you said, there is really not much folks can do about this right now. (In [my PR on the docs](https://github.com/salesforce-experience-platform/lwc-docs/pull/246), I advise to use light DOM right now. But that's not always feasible obviously.)

Reply:
In that case, can we do one or more of the following for this PR?

1) coalesce based on tagName and log once (or add a TODO for this)
2) put the warnings on line 60 behind some `verbose` reporting flag
2) get rid of the console warnings on line 60, but delegate logging to the console to the dispatcher itself (may need to add ability to pass ReportingEventId-specific parameters to the dispatcher).

Reply:
Done: 3958e4c5c